feat(vitals): Add an endpoint for the new vitals cards

wmak opened the pull request:
- This introduces an endpoint for vital cards that get the counts of
  each vital across their thresholds, along with their p75 and total
  count

wmak committed:
feat(vitals): Add an endpoint for the new vitals cards

- This introduces an endpoint for vital cards that get the counts of
  each vital across their thresholds, along with their p75 and total
  count

k-fish commented on tests/snuba/api/endpoints/test_organization_events_vitals.py:
```diff
@@ -0,0 +1,164 @@
+        self.query.update({"vital": ["measurements.lcp", "measurements.fcp"]})
+        response = self.do_request()
+        assert response.status_code == 200
+        assert response.data["measurements.lcp"] == {
```
Should we test all of them here since we are partially testing in the other tests?

wmak replied:
Ah yeah :+1: makes sense

k-fish commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+
+class OrganizationEventsVitalsEndpoint(OrganizationEventsV2EndpointBase):
+    VITALS = {
+        "measurements.lcp": {"thresholds": [0, 2500, 4000]},
```
Are we hardcoding these to just web vitals? Do we need groups if that's the case? I highly doubt this will be extended beyond the 3 thresholds in the next year or two.

markstory replied:
When we get around to custom measures, we could have thresholds be configurable via settings. I think this will work for now, and could extend into backend measures when they get added.

Zylphrex commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+            )
+
+        results = {}
+        event_data = events_results["data"][0]
```
Can we add a check to make sure there is exactly 1 row in the results? If there is a hiccup in snuba, we wouldn't want sentry to error too.

Zylphrex commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+            # Go backwards so that we can subtract and get totals
+            for i in range(groups - 1, -1, -1):
+                count = event_data[aliases[vital][i]]
+                if count:
+                    group_count = count - total
+                else:
+                    group_count = None
+                results[vital][f"group_{i}"] = group_count
+                if group_count is not None:
+                    total += group_count
```
Would using `0`'s here instead of `None` make more sense so we can get only numeric results? As it stands, it's possible to see a mix of numbers and `None` in the results.

```suggestion
                group_count = count - total if count else 0
                results[vital][f"group_{i}"] = group_count
                total += group_count
```

markstory replied:
👍 for using 0 instead of None.

wmak replied:
Hmm 0 is simpler too so I don't mind switching, but my reason for this was that `p75` would already be None in this case so I wanted to be consistent.

Zylphrex replied:
I think using `None` for `p75` makes sense as that would represent the case that there are no measurements available. But for the count, `0` does a better job of representing the no measurement case.

wmak replied:
Updated in 6578457

Zylphrex commented on tests/snuba/api/endpoints/test_organization_events_vitals.py:
```diff
@@ -0,0 +1,164 @@
+        self.query.update({"vital": ["measurements.lcp", "measurements.fcp"]})
+        response = self.do_request()
+        assert response.status_code == 200
+        assert response.data["measurements.lcp"] == {
+            "group_0": None,
+            "group_1": None,
+            "group_2": None,
+            "total": 0,
+            "p75": None,
+        }
+        assert response.data["measurements.fcp"] == {
+            "group_0": None,
+            "group_1": None,
+            "group_2": None,
+            "total": 0,
+            "p75": None,
+        }
```
See above comment about `0`'s vs `None`.

markstory commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+                    raise ParseError(detail=u"{} is not a valid vital".format(vital))
+                aliases[vital] = []
+                for index, threshold in enumerate(self.VITALS[vital]["thresholds"]):
+                    column = f"count_at_least({vital}, {threshold})"
```
Would having this validation and conversion code live in a serializer help make it simpler?

wmak replied:
I think for now I'd rather keep this here cause its only validating that the vitals are real. But if there's a need to validate more I think it'd be worth moving this to a serializer.

markstory commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+                selected_columns=selected_columns,
+                query=request.GET.get("query"),
+                params=params,
+                # Results should only ever have 1 page
```
```suggestion
                # Results should only ever have 1 result
```

markstory commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+                    group_count = count - total
+                else:
+                    group_count = None
+                results[vital][f"group_{i}"] = group_count
```
This will 💣 the py2 builds.

wmak replied:
🙃 yeah I added these thinking we'd be py3 ready by the time I PR-ed this, then forgot to remove, will remove 😢

Zylphrex commented on src/sentry/api/endpoints/organization_events_vitals.py:
```diff
@@ -0,0 +1,80 @@
+        for vital in vitals:
+            groups = len(aliases[vital])
+            results[vital] = {}
+            total = 0
```
Doesn't `total` always equal to `count_at_least(..., 0)`? In fact, as long as you're assuming that thresholds array is in ascending order (which you are), it's always equal to `count_at_least(..., smallest thresholds)`

wmak replied:
It does, but we'll still need to keep a running total to calculate the first n-1 thresholds (though tbf, in the foreseeable future n=3). I could set total via `results[vital]["total"] = event_data[aliases[vital][0]]`, but IMO this is more readable than that

wmak committed:
ref: Addressing PR comments

- Adding tests with all the vitals instead of only a few
- Add a check that there's exactly one result to avoid erroring when
  there aren't any
  - I don't think there's a need to check for >1 since that means snuba
    limiting isn't working anymore
- Showing 0 instead of None when measurements don't exist
- Removing f-strings :(

markstory commented on tests/snuba/api/endpoints/test_organization_events_vitals.py:
```diff
@@ -0,0 +1,195 @@
+        self.query.update({"vital": ["measurements.lcp"]})
+        response = self.do_request()
+        assert response.status_code == 200
+        print(response.data)
```
```suggestion

```

Zylphrex reviewed (approved):
LGTM

wmak committed:
Update tests/snuba/api/endpoints/test_organization_events_vitals.py

Co-authored-by: Mark Story <mark@sentry.io>

wmak committed:
Merge branch 'master' into feat/vitals-endpoint

wmak committed:
ref: Extra import from merge conflict

wmak committed:
Merge branch 'master' into feat/vitals-endpoint
